core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -114,6 +120,20 @@
     private final AtomicInteger throttleRequestCount = new AtomicInteger();
     private final EngineConfig.OpenMode openMode;
     private final AtomicBoolean allowCommits = new AtomicBoolean(true);
+    /**
+     * Refresh listeners. While they are not stored in sorted order they are processed as though they are.
+     */
+    private final LinkedTransferQueue<RefreshListener> refreshListeners = new LinkedTransferQueue<>();
```

Comment:
ShadowEngine has a ShadowIndexShard which can just override the method. This stuff especially the threadpool execution does not belong in the engine.